File: data-structures/linked-list/singly_linked_list.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Generic, Optional, TypeVar, List

T = TypeVar("T")

@dataclass
class ListNode(Generic[T]):
    value:T
    next:Optional["ListNode[T]"] = None


class SinglyLinkedList(Generic[T]):
    def __init__(self):
        self.head: Optional[ListNode[T]] = None
        self.tail: Optional[ListNode[T]] = None
        self.length: int = 0
    
    def prepend(self,value):
        node = ListNode(value)

        if self.head is None:
            self.head = self.tail = node
        else:
            node.next = self.head
            self.head = node
        self.length += 1

    def append(self,value):
        node = ListNode(value)
        if self.head is None:
            self.head = self.tail = node
        else:
            self.tail.next = node
            self.tail = node
        self.length+=1
    
    def get(self,index:int):
        if index < 0 or index >= self.length:
            return None
        cur = self.head
        while(index>0):
            cur = cur.next
            index-=1
        return cur
    
    def list_to_arr(self)->List[T]:
        result:List[T] = []
        cur = self.head
        while cur is not None:
            result.append(cur.value)
            cur = cur.next
        
        return result
    
    def remove_head(self):
        if self.head is None:
            return None
        
        removed = self.head
        self.head = removed.next
        self.length -= 1

        if self.length == 0:
            self.head = self.tail = None
        
        removed.next = None
        return removed
    
    def remove_tail(self) -> Optional[ListNode[T]]:
        if self.head is None:
            return None

        # one node
        if self.length == 1:
            removed = self.head
            self.head = self.tail = None
            self.length = 0
            return removed

        cur = self.head
        assert self.tail is not None

        while cur.next is not self.tail:
            cur = cur.next  # type: ignore[assignment]

        removed = self.tail
        cur.next = None
        self.tail = cur
        self.length -= 1

        removed.next = None
        return removed
    
    def reverse(self):
        if self.head is None:
            return None
        cur = self.head
        prev = None
        self.tail = self.head
        while cur is not None:
            nxt = cur.next
            cur.next = prev
            prev = cur
            cur = nxt
        self.head = prev


ll = SinglyLinkedList[int]()
ll.append(10)
ll.append(20)
ll.prepend(5)
print(ll.list_to_arr())  # [5, 10, 20]

ll.remove_tail()
print(ll.list_to_arr())  # [5, 10]

ll.reverse()
print(ll.list_to_arr())  # [10, 5]
print(ll.length)  

File: data-structures/linked-list/singly-linked-list.ts
class ListNode<T> {
  value: T;
  next: ListNode<T> | null;

  constructor(value: T, next: ListNode<T> | null = null) {
    this.value = value;
    this.next = next;
  }
}

class SignlyLinkedList<T> {
  head: ListNode<T> | null = null;
  tail: ListNode<T> | null = null;
  length: number = 0;

  prepend(value: T): void {
    const node = new ListNode(value);
    if (!this.head) {
      this.head = node;
      this.tail = node;
    } else {
      node.next = this.head;
      this.head = node;
    }
    this.length++;
  }

  append(value: T): void {
    const node = new ListNode(value);
    if (!this.head) {
      this.prepend(value);
      return;
    } else {
      this.tail!.next = node;
      this.tail = node;
    }
    this.length++;
  }

  removeHead(): ListNode<T> | null {
    if (!this.head) return null;
    const removed = this.head;
    this.head = removed.next;
    this.length--;
    if (this.length === 0) {
      this.head = this.tail = null;
    }
    removed.next = null;
    return removed;
  }

  removeTail(): ListNode<T> | null {
    if (!this.head) return null;
    if (this.length === 1) {
      const removed = this.head;
      this.head = this.tail = null;
      this.length = 0;
      return removed;
    }

    let curr = this.head;
    while (curr.next !== this.tail) {
      curr = curr.next!;
    }
    const removed = this.tail;
    curr.next = null;
    this.tail = curr;
    this.length--;
    return removed;
  }

  get(index: number): ListNode<T> | null {
    if (index < 0 || index >= this.length) return null;
    if (index == 0) return this.head;
    if (index === this.length - 1) return this.tail;
    let curr = this.head;
    while (index > 0) {
      curr = curr!.next;
      index--;
    }
    return curr;
  }

  removeAt(index: number): ListNode<T> | null {
    if (index < 0 || index >= this.length) return null;
    if (index === 0) return this.removeHead();
    if (index === this.length - 1) return this.removeTail();
    const prevNode = this.get(index - 1);
    if (!prevNode) return null;
    const removeNode = prevNode.next;
    if (!removeNode) return null;
    prevNode.next = removeNode?.next;
    this.length--;
    return removeNode;
  }

  insertAt(index: number, value: T): void {
    if (index < 0 || index > this.length) return;
    if (index === 0) return this.prepend(value);
    if (index === this.length - 1) return this.append(value);
    const node = new ListNode(value);
    const prevNode = this.get(index - 1);
    if (!prevNode) return;
    const currNode = prevNode.next;
    prevNode.next = node;
    node.next = currNode;
    this.length++;
  }

  listToArray(): T[] {
    const result: T[] = [];
    let curr = this.head;
    while (curr !== null) {
      result.push(curr!.value);
      curr = curr!.next;
    }
    return result;
  }

  reverse(): void {
    if (!this.head || this.length === 1) return;
    let prev: ListNode<T> | null = null;
    let cur: ListNode<T> | null = this.head;
    this.tail = this.head;
    while (cur !== null) {
      const next: ListNode<T> | null = cur.next;
      cur.next = prev;
      prev = cur;
      cur = next;
    }
    this.head = prev;
  }
}

const list = new SignlyLinkedList();
list.append(10);
list.prepend(20);
list.prepend(30);
list.prepend(40);
list.prepend(50);
list.prepend(60);
list.insertAt(5, 75);
console.log(list.removeAt(3));
console.log(list.length);
